fem/libceed/coefficient.hpp

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_LIBCEED_COEFF
+#define MFEM_LIBCEED_COEFF
+
+#include "../../config/config.hpp"
+#ifdef MFEM_USE_CEED
+#include "util.hpp"
+#include "../../linalg/vector.hpp"
+#include "../../linalg/dtensor.hpp"
+#include "../../mesh/mesh.hpp"
+#include "../gridfunc.hpp"
+#include "../../linalg/dtensor.hpp"
+
+namespace mfem
+{
+
+class Mesh;
+class IntegrationRule;
+class Coefficient;
+class VectorCoefficient;
+class GridFunction;
+
+namespace ceed
+{
+
+struct Coefficient
+{
+   const int ncomp;
+   Coefficient(int ncomp_) : ncomp(ncomp_) { }
+   virtual bool IsConstant() const { return true; }
+   virtual ~Coefficient() { }
+};
+
+struct VariableCoefficient : Coefficient
+{
+   CeedVector coeffVector = nullptr;
+   const CeedEvalMode emode;
+   VariableCoefficient(int ncomp_, CeedEvalMode emode_)
+      : Coefficient(ncomp_), emode(emode_) { }
+   virtual bool IsConstant() const override { return false; }
+   ~VariableCoefficient()
+   {
+      CeedVectorDestroy(&coeffVector);
+   }
+};
+
+struct GridCoefficient : VariableCoefficient
+{
+   const mfem::GridFunction &gf;
+   CeedBasis basis;
+   CeedElemRestriction restr;
+   GridCoefficient(const mfem::GridFunction &gf_)
+      : VariableCoefficient(gf_.VectorDim(), CEED_EVAL_INTERP),
+        gf(gf_)
+   {
+      InitVector(gf, coeffVector);
+   }
+};
+
+struct QuadCoefficient : VariableCoefficient
+{
+   mfem::Vector coeff;
+   CeedElemRestriction restr;
+   QuadCoefficient(int ncomp_) : VariableCoefficient(ncomp_, CEED_EVAL_NONE) { }
+};
+
+/** @brief Identifies the type of coefficient of the Integrator to initialize
```

Comment:
This documentation could be improved. The reference to `Integrator` is confusing because that type isn't used at all here. The documentation should specify which of `Q` and `coeff_ptr` is input and which is output, and probably emphasize that `coeff_ptr` is `mfem::ceed::Coefficient`, a different type from `Q`.

Reply:
I improved the documentation, tell me if you like it.